Thread-safety

ashwin-hari opened the pull request:
Addresses thread-safety, issue #2 

* caller can provide function pointers for lock and unlock
* Example scenario of how `lock` and `unlock` may be used: An application stores logs; each log's owner field is set to a value that identifies which component of the application produced the log. Threads are used to read and process logs for a specific owner. The number of threads would be less than the number of components in the application. Latency doesn't matter here.
* a basic multi-threaded test
* validated locally using ctest, `gcc-8`, Ubuntu 20.04

fefa4ka commented on src/lr.c:
```diff
@@ -1,24 +1,39 @@
 #include "lr.h"
 #include <stdio.h>
 
+#define RETURN_IF_ERROR(x) do { \
```
I know this is not common, but the library already uses lower case for macros (moreover, for internal use), what do you think about `lr_assert`?

ashwin-hari replied:
sure i can use lower case. I don't like `lr_assert` since it is similar in name to the `assert` library function, calls `abort` and terminates the program. 

what do you think of `return_if_error`

fefa4ka replied:
`return_if_error` looks good

ashwin-hari replied:
I had to add some different macros and remove this one, but I've followed the lower snake case convention.

fefa4ka commented on include/lr.h:
```diff
@@ -47,6 +47,25 @@ struct lr_cell {
+    /* Blocks until the mutex is acquired.
+     * Returns LR_OK on success and an error code otherwise. 
+     */
+    enum lr_result (*lock)();
```
maybe you have some ideas about the basic implementation of a `lock()` and `unlock()`  without hardware support in regular C for bare metal?

fefa4ka replied:
Looks like there are some hidden pitfalls in cases without hardware support for such things. Maybe provide argument for some flexibility?

fefa4ka replied:
Something like this maybe
```c
void mutex_lock(struct linked_ring *lr, lr_owner_t owner)
{
    // Busy-wait until the mutex is not locked
    while (lr->locked || lr->locked != owner) { }
    
    // Atomically set the mutex to locked
    lr->locked = owner;
}

void mutex_unlock(struct linked_ring *lr)
{
    // Atomically set the mutex to unlocked
    lr->locked = 0;
}
```

fefa4ka replied:
In this case, we have to provide an additional pointer to the lock storage

ashwin-hari replied:
yes, i think having an argument would help. How about this?

```
struct mutex_attr
{
  void *state;
  enum lr_result (*lock)(void *state, lr_owner_t owner);
  enum lr_result (*unlock)(void *state);
};

lr_mutex_init(mutex_attr *attr);
```

In your example, we would have

```
struct mutex_state
{
    int *atomic_mutex_reg = 0xFFFF;
};

void mutex_lock(void *state, lr_owner_t owner)
{
    mutex_state *s = (mutex_state *) state;
    // Busy-wait until the mutex is not locked
    while ( *(s->atomic_mutex_reg) == 0 || *(s->atomic_mutex_reg) != owner) { }
    
    // Atomically set the mutex to locked
    *(s->atomic_mutex_reg) = owner;
}

void mutex_unlock(void *state)
{
    // Atomically set the mutex to unlocked
    mutex_state *s = (mutex_state *) state;
    *(s->atomic_mutex_reg) = owner;
}
```

fefa4ka replied:
maybe `atomic_mutex_reg` should have the same type as `owner`, because bit wide is platform dependent. by default type is `uintptr_t`. In case of arbitrary bits wide, looks like using of default 0 value makes sense for clear and understandable initialization.

ashwin-hari replied:
thanks for the in-depth answers. I've updated the `lock` and `unlock` interfaces - please reopen or  add more comments if you see any issues.

fefa4ka commented on include/lr.h:
```diff
@@ -55,6 +74,8 @@ struct linked_ring {
     struct lr_cell *read;  // Cell that is currently being read from
     struct lr_cell *write; // Cell that is currently being written to
+
+    struct lr_recursive_mutex mutex; // used to make operations thread-safe
```
Maybe enough to add just two methods to this structure, short `lr->unlock()` calls looks good. What do you think? I see self explanatory structure name that suggest to implement right type of lock, generally, I like this naming.

ashwin-hari replied:
i like the short `lr->unlock`; I'll make the change.

ashwin-hari replied:
I've made the change.

ashwin-hari committed:
Add recursive mutex

ashwin-hari commented:
test

fefa4ka commented on test/edge.c:
```diff
@@ -37,14 +37,25 @@ typedef enum {
 } owner_t;
 
+// used to mock lock and unlock for single-threaded tests
+enum lr_result always_return_ok() 
```
Maybe better approach is check `lock` and `unlock` pointers inside library methods. I don't know how optimisation work in this case, but look like method calling and returning is slower that check for `NULL`.

ashwin-hari replied:
yes, sure. that makes sense.

ashwin-hari replied:
I've made the change.

ashwin-hari committed:
pr feedback

fefa4ka commented:
Looks great! Wood be nice to change README, maybe add some info about thread-safety feature, and modify contribution section, at least no need inspire to do blocking mechanics.

fefa4ka reviewed (commented):
looks great, thanks! with built-in example for bare-metal blocking would look even more gorgeous.

ashwin-hari committed:
pr feedback 2

ashwin-hari commented:
@fefa4ka thanks, I've pushed another iteration

* updated readme 
* added an example using atomics
* updated the interface of `unlock` to also take an owner argument; this allows the implementation to perform more error-checking like trying to unlock a mutex acquired by another thread, etc.
* removed the requirement that `lock` be recursive 

could you also build on your platform and ensure things work for you? I verified on my machine using Ubuntu 20.04 and gcc 8.4.0

fefa4ka commented:
OS: Mac OS 13.1
Compiler: Apple clang version 14.0.0 (clang-1400.0.29.202) 
Compiler: gcc-12 (Homebrew GCC 12.2.0) 12.2.0
Target: arm64-apple-darwin22.2.0

```
❯ ./test_multi_thread
OK      Failed to initialize buffer
OK      Failed to create thread
OK      Failed to create thread
OK      Failed to create thread
OK      Failed to create thread
ERROR   Error while getting data from buffer (/Users/fefa4ka/Documents/linked_ring/test/multi_thread.c:127)

ERROR   Data does not match (/Users/fefa4ka/Documents/linked_ring/test/multi_thread.c:132)

ERROR   Test failed (/Users/fefa4ka/Documents/linked_ring/test/multi_thread.c:214)
```

@ashwin-hari I will try to debug

fefa4ka commented:
```
(lldb) f 0
frame #0: 0x0000000100002bd0 test_multi_thread`test_multiple_threads(num_threads=2, func=(test_multi_thread`put_get_data at multi_thread.c:115:1)) at multi_thread.c:170:9
   167      {
   168          owners[i] = i;
   169          int ret = pthread_create(&threads[i], NULL, func, (void *) &owners[i]);
-> 170          test_assert(ret == 0, "Failed to create thread");
   171      }
   172
   173      enum lr_result result = LR_OK;
   
(lldb) v
OK      Failed to initialize buffer
(unsigned int) num_threads = 2
(void *(*)(void *)) func = 0x0000000100002870 (test_multi_thread`put_get_data at multi_thread.c:115:1)
(unsigned int) buffer_size = 1
(pthread_t[4]) threads = {
  [0] = 0x000000016fe87000
  [1] = NULL
  [2] = 0x000000016fdfef70
  [3] = 0x0000000100002e18
}
(unsigned int[4]) owners = ([0] = 0, [1] = 1, [2] = 10744, [3] = 236519425)
(pthread_mutexattr_t) pthread_attr = (__sig = 1297373249, __opaque = "\0\0\0\0\xa4\0\0")
(lr_mutex_attr) attr = {
  state = 0x0000000100008040
  lock = 0x0000000100002670 (test_multi_thread`pthread_lock at multi_thread.c:35:1)
  unlock = 0x00000001000026ac (test_multi_thread`pthread_unlock at multi_thread.c:48:1)
}
(lr_result) result = 0x10100
(unsigned int) i = 0
(int) ret = 0
```
